File: app/src/main/java/linusfessler/alarmtiles/timer/Timer.java
package linusfessler.alarmtiles.timer;

import androidx.room.Entity;
import androidx.room.PrimaryKey;

import lombok.Data;

@Data
@Entity
public class Timer {

    @PrimaryKey(autoGenerate = true)
    private Long id;

    private boolean enabled;
    private long duration;
    private Long startTimeStamp;

    // TODO: Add a constructor with sensible default values
    public Timer() {
        this.duration = 10000;
    }

    void toggle() {
        if (this.isEnabled()) {
            this.disable();
        } else {
            this.enable();
        }
    }

    void enable() {
        this.setEnabled(true);
        this.setStartTimeStamp(System.currentTimeMillis());
    }

    void disable() {
        this.setEnabled(false);
        this.setStartTimeStamp(null);
    }
}

File: app/src/main/java/linusfessler/alarmtiles/timer/Timer.kt
package linusfessler.alarmtiles.timer

import androidx.room.Entity
import androidx.room.PrimaryKey

@Entity
data class Timer(
        @field:PrimaryKey val id: Long = 0,
        val isEnabled: Boolean = false,
        val duration: Long = 10000,
        val startTimeStamp: Long? = null
) {
    fun toggle(): Timer {
        return if (isEnabled)
            disable()
        else
            enable()
    }

    private fun enable(): Timer {
        return copy(isEnabled = true, startTimeStamp = System.currentTimeMillis())
    }

    fun disable(): Timer {
        return copy(isEnabled = false, startTimeStamp = null)
    }
}